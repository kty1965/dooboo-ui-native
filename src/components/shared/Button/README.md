# Button

- [![Npm Version](http://img.shields.io/npm/v/@dooboo-ui/native-button.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-button)
- [![Downloads](http://img.shields.io/npm/dm/@dooboo-ui/native-button.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-button)


> [Button] component that can be used inside product. Has basic features like `loading` state, `disabled` state and also has ability to put `img` to left-hand which is used very often.

![image](https://user-images.githubusercontent.com/27461460/62291727-9be84100-b49f-11e9-8ce5-ceaa1dc3153e.png)


## Props

|                | necessary | types                | default |
| -------------- | --------- | -------------------- | ------- |
| testID         |           | string               |         |
| style          |           | ViewStyle            |         |
| textStyle      |           | TextStyle            |         |
| dark           |           | boolean              |         |
| inverted       |           | boolean              |         |
| isLoading      |           | boolean              |         |
| isDisabled     |           | boolean              |         |
| iconLeft       |           | ReactElement         |         |
| iconRight      |           | ReactElement         |         |
| indicatorColor |           | string               |'#ffffff'|
| activeOpacity  |           | number               |   0.5   |
| children       |           | string, ReactElement |         |
| text           |           | string               |         |
| onClick        |           | func                 |         |

## Installation

```sh
yarn add @dooboo-ui/native
```

or

```sh
yarn add @dooboo-ui/native-button
```

## Getting started

- Import

  ```javascript
  import { Button } from '@dooboo-ui/native';
  ```

- Usage
  - with `StyleSheet`
  ```jsx
  <Button
    testID="sampleButton"
    style={{
      backgroundColor: 'red',
      borderColor: 'blue',
    }}
    textStyle={{
      color: 'white',
    }}
    onClick={(): void => {}}
  >
    Sample button
  </Button>
  ```
  ```jsx
  <Button
    testID="sampleButton"
    style={{
      backgroundColor: 'red',
      borderColor: 'blue',
    }}
    onClick={(): void => {}}
  >
    <SampleText>Sample button</SampleText>
  </Button>
  ```

  - with `styled-components`
  ```jsx
  const SampleButton = styled(Button)`
    width: 136px;
    height: 60px;
    background-color: black;
    border-color: red;
    border-width: 1px;
  `;
  const SampleText = styled.Text`
    color: white;
  `;
  <SampleButton
    testID="sampleButton"
    textStyle={{
      color: 'white',
    }}
    onClick={(): void => {}}
  >
    Sample button
  </SampleButton>
  ```
  ```jsx
  const SampleButton = styled(Button)`
    width: 136px;
    height: 60px;
    background-color: black;
    border-color: red;
    border-width: 1px;
  `;
  const SampleText = styled.Text`
    color: white;
  `;
  <SampleButton
    testID="sampleButton"
    onClick={(): void => {}}
  >
    <SampleText>Sample button</SampleText>
  </SampleButton>
  ```

  - example
  ```jsx
  function Page(props: Props) {
    return (
      <Container>
        <Button
          testID="btn"
          isLoading={false}
          onClick={() => {}}
        >
          üòÄ üòé üëç üíØ
        </Button>
        <Button
          style={{
            marginVertical: 40,
          }}
          isDisabled={true}
          onClick={() => {}}
        >
          This is disabled!!
        </Button>
        <Button
          testID="btnGoogle"
          iconLeft={<Image source={IC_GOOGLE} />}
          isLoading={googleLoading}
          indicatorColor="#023059"
          onClick={() => {
            setGoogleLoading(true);
            const timeout = setTimeout(() => {
              setGoogleLoading(false);
              clearTimeout(timeout);
            }, 2000);
          }}
        >
          GOOGLE SIGN IN
        </Button>
        <Button
          testID="btnFacebook"
          iconLeft={<Image source={IC_FACEBOOK} />}
          indicatorColor="#023059"
          isLoading={facebookLoading}
          style={{
            marginTop: 40,
            backgroundColor: '#ccc',
            borderWidth: 0.5,
            borderRadius: 0,
          }}
          onClick={() => {
            setFacebookLoading(true);
            const timeout = setTimeout(() => {
              setFacebookLoading(false);
              clearTimeout(timeout);
            }, 2000);
          }}
        >
          FACEBOOK SIGN IN
        </Button>
      </Container>
    );
  }
  ```