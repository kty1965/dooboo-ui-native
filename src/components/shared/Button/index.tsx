import {
  ActivityIndicator,
  GestureResponderEvent,
  StyleProp,
  StyleSheet,
  TextStyle,
  TouchableOpacity,
  ViewStyle,
} from 'react-native';
import styled, { DefaultTheme, ThemeProps } from 'styled-components/native';

import React from 'react';

interface TextThemeType {
  fontColor: string;
}

interface ButtonThemeType {
  backgroundColor: string;
  borderColor: string;
}

interface ButtonType extends ThemeProps<ButtonThemeType> {
  theme: ButtonThemeType;
}

interface TextType extends ThemeProps<TextThemeType>{
  theme: TextThemeType;
}

interface ThemeType extends DefaultTheme {
  backgroundColor: string;
  borderColor: string;
  fontColor: string;
}

interface StatefulThemeType extends ThemeType {
  INVERTED: ThemeType;
  DISABLED: ThemeType;
}

interface Props {
  testID?: string;
  style: StyleProp<ViewStyle>;
  textStyle: StyleProp<TextStyle>;
  dark?: boolean;
  inverted?: boolean;
  isLoading?: boolean;
  isDisabled?: boolean;
  iconLeft?: React.ReactElement;
  iconRight?: React.ReactElement;
  indicatorColor: string;
  activeOpacity: number;
  children?: string | React.ReactElement;
  text?: string;
  onClick?: (e?: GestureResponderEvent) => void | Promise<void>;
}

export const TESTID = {
  BUTTON: 'button',
  ACTIVITYINDICATOR: 'activityIndicator',
  ICONLEFT: 'iconLeft',
  ICONRIGHT: 'iconRight',
  TEXT: 'text',
};

const COLOR: {
  [key: string]: string;
} = {
  WHITE: '#ffffff',
  DODGERBLUE: '#3a8bff',
  VERYLIGHTGRAY: '#cccccc',
  LIGHTGRAY: '#c8c8c8',
  BLUE: '#0000ff',
  STRONGBLUE: '#069ccd',
  GRAY3: '#080808',
  GRAY7: '#121212',
  GRAY59: '#969696',
};

export const THEME: {
  LIGHT: StatefulThemeType;
  DARK: StatefulThemeType;
} = {
  LIGHT: {
    backgroundColor: COLOR.WHITE,
    borderColor: COLOR.BLUE,
    fontColor: COLOR.STRONGBLUE,
    INVERTED: {
      backgroundColor: COLOR.BLUE,
      borderColor: COLOR.STRONGBLUE,
      fontColor: COLOR.WHITE,
    },
    DISABLED: {
      backgroundColor: COLOR.VERYLIGHTGRAY,
      borderColor: COLOR.LIGHTGRAY,
      fontColor: COLOR.GRAY59,
    },
  },
  DARK: {
    backgroundColor: COLOR.WHITE,
    borderColor: COLOR.GRAY7,
    fontColor: COLOR.GRAY3,
    INVERTED: {
      backgroundColor: COLOR.GRAY7,
      borderColor: COLOR.GRAY3,
      fontColor: COLOR.WHITE,
    },
    DISABLED: {
      backgroundColor: COLOR.VERYLIGHTGRAY,
      borderColor: COLOR.LIGHTGRAY,
      fontColor: COLOR.GRAY59,
    },
  },
};

const StyledButton = styled.View<ButtonType>`
  width: 320px;
  height: 52px;
  background-color: ${({ theme }): string => theme.backgroundColor};
  border-color: ${({ theme }): string => theme.borderColor};
  border-radius: 4px;
  border-width: 2px;
  justify-content: center;
  align-items: center;
  flex-direction: row;
`;

const Text = styled.Text<TextType>`
  font-size: 14px;
  color: ${({ theme }): string => theme.fontColor};
`;

const IconLeft = styled.View`
  position: absolute;
  left: 0;
`;

const IconRight = styled.View`
  position: absolute;
  right: 0;
`;

const getDefaultTheme = ({
  dark,
  inverted,
  isDisabled,
}: {
  dark?: boolean;
  inverted?: boolean;
  isDisabled?: boolean;
}): ThemeType => {
  const theme = dark ? THEME.DARK : THEME.LIGHT;

  if (isDisabled) {
    return theme.DISABLED;
  }

  return inverted ? theme.INVERTED : theme;
};

const getText = ({
  children,
  text,
  style,
  theme,
}: {
  children?: string | React.ReactElement;
  text?: string;
  style: TextStyle;
  theme: TextThemeType;
}): string | React.ReactElement | undefined => {
  if (typeof children === 'undefined') {
    return (
      <Text testID={TESTID.TEXT} style={style} theme={theme}>{text}</Text>
    );
  }

  if (typeof children === 'string') {
    return (
      <Text testID={TESTID.TEXT} style={style} theme={theme}>{children}</Text>
    );
  }

  return children;
};

function Button(props: Props): React.ReactElement {
  const {
    testID,
    style,
    textStyle,
    dark,
    inverted,
    isLoading,
    isDisabled,
    iconLeft,
    iconRight,
    indicatorColor,
    activeOpacity,
    children,
    text,
    onClick,
  } = props;

  const buttonStyleToApply = StyleSheet.flatten(style);
  const textStyleToApply = StyleSheet.flatten(textStyle);

  const defaultTheme = getDefaultTheme({
    dark,
    inverted,
    isDisabled,
  });

  const buttonThemeToApply: ButtonThemeType = {
    backgroundColor: buttonStyleToApply.backgroundColor || defaultTheme.backgroundColor,
    borderColor: buttonStyleToApply.borderColor || defaultTheme.borderColor,
  };

  const textThemeToApply: TextThemeType = {
    fontColor: textStyleToApply.color || defaultTheme.fontColor,
  };

  const textToRender = getText({
    children,
    text,
    style: textStyleToApply,
    theme: textThemeToApply,
  });

  return (
    <TouchableOpacity
      testID={testID}
      activeOpacity={activeOpacity}
      onPress={onClick}
      disabled={isDisabled}
    >
      <StyledButton
        style={buttonStyleToApply}
        theme={buttonThemeToApply}
      >
        {!isLoading && <IconLeft testID={TESTID.ICONLEFT}>{iconLeft}</IconLeft>}
        {!isLoading && textToRender}
        {!isLoading && <IconRight testID={TESTID.ICONRIGHT}>{iconRight}</IconRight>}
        {isLoading && <ActivityIndicator testID="activityIndicator" size="small" color={indicatorColor} />}
      </StyledButton>
    </TouchableOpacity>
  );
}

Button.defaultProps = {
  style: {},
  textStyle: {},
  indicatorColor: COLOR.WHITE,
  activeOpacity: 0.5,
};

export default Button;
