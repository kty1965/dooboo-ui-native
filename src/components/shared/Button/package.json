{
  "name": "@dooboo-ui/native-button",
  "main": "package/index.js",
  "types": "package/index.d.ts",
  "version": "0.4.0",
  "license": "MIT",
  "scripts": {
    "build": "rm -rf ./package && tsc",
    "watch": "tsc -w"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/dooboolab/dooboo-ui-native.git",
    "directory": "src/components/shared/Button"
  },
  "bugs": {
    "url": "https://github.com/dooboolab/dooboo-ui-native/issues"
  },
  "homepage": "https://github.com/dooboolab/dooboo-ui-native#readme",
  "dependencies": {
    "dooboolab-welcome": "^1.1.1"
  },
  "devDependencies": {
    "styled-components": "^4.4.0",
    "typescript": "^3.6.4"
  },
  "peerDependencies": {
    "react": ">=16.8",
    "react-native": ">=0.59",
    "styled-components": ">=4.4.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "button",
    "react-hook",
    "react-native",
    "dooboo-ui/native",
    "dooboolab"
  ]
}
