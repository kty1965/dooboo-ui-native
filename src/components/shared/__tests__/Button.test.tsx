import * as React from 'react';

import Button, { TESTID, THEME } from '../Button';
import {
  NativeTestInstance,
  NativeTestInstanceJSON,
  RenderResult,
  act,
  fireEvent,
  render,
} from '@testing-library/react-native';
import { StyleProp, Text, TextStyle, ViewStyle } from 'react-native';

import { ReactTestInstance } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';

interface Props {
  testID?: string;
  style?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  dark?: boolean;
  inverted?: boolean;
  isLoading?: boolean;
  isDisabled?: boolean;
  iconLeft?: React.ReactElement;
  iconRight?: React.ReactElement;
  indicatorColor?: string;
  activeOpacity?: number;
  children?: string | React.ReactElement;
  text?: string;
  onClick?: (params?: any) => void | Promise<void>;
}

const renderComponent = (props?: Props): RenderResult => {
  const defaultProps = {
    testID: TESTID.BUTTON,
  };

  const propsToApply = props ? {
    ...defaultProps,
    ...props,
  } : defaultProps;

  const component = (
    <ThemeProvider theme={THEME.LIGHT}>
      <Button {...propsToApply} />
    </ThemeProvider>
  );

  return render(component);
};
const renderComponentAsJson = (props?: Props): NativeTestInstanceJSON => renderComponent(props).asJSON();
const getButton = (comp: RenderResult): NativeTestInstance | null => comp.getByTestId(TESTID.BUTTON);
const getText = (comp: RenderResult): NativeTestInstance | null => comp.getByTestId(TESTID.TEXT);
const queryIconLeft = (comp: RenderResult): NativeTestInstance | null => comp.queryByTestId(TESTID.ICONLEFT);
const queryIconRight = (comp: RenderResult): NativeTestInstance | null => comp.queryByTestId(TESTID.ICONRIGHT);
const queryActivityIndicator =
  (comp: RenderResult): NativeTestInstance | null => comp.queryByTestId(TESTID.ACTIVITYINDICATOR);

describe('[Button]', () => {
  it('renders without crashing', () => {
    const comp = renderComponentAsJson();

    expect(comp).toMatchSnapshot();
    expect(comp).toBeTruthy();
  });

  it('renders without crashing when children is not empty', () => {
    const comp = renderComponentAsJson({
      children: (
        <Text>Sample text</Text>
      ),
    });

    expect(comp).toMatchSnapshot();
    expect(comp).toBeTruthy();
  });

  it('should render text by the value of text prop', async () => {
    const SAMPLE_TEXT = 'sampleText';
    const comp = renderComponent({
      text: SAMPLE_TEXT,
    });

    const text = await getText(comp);
    expect(text.children).toContain(SAMPLE_TEXT);
  });

  it('should render children', async () => {
    const SAMPLE_ELEMENT_TESTID = 'sampleElement';
    const SAMPLE_TEXT = 'sampleElementText';
    const sampleElement = (
      <Text testID={SAMPLE_ELEMENT_TESTID}>{SAMPLE_TEXT}</Text>
    );
    const comp = renderComponent({
      children: sampleElement,
    });

    const text = await comp.queryByTestId(SAMPLE_ELEMENT_TESTID);
    expect(text.children).toContain(SAMPLE_TEXT);
  });

  it('should render text when type of children is string', async () => {
    const SAMPLE_TEXT = 'sampleElementText';
    const comp = renderComponent({
      children: SAMPLE_TEXT,
    });

    const text = await getText(comp);
    expect(text.children).toContain(SAMPLE_TEXT);
  });

  it('should not render [ActivityIndicator] as default', async () => {
    const comp = renderComponent();
    const activityIndicator = await queryActivityIndicator(comp);

    expect(activityIndicator).toBeNull();
  });

  it('should render [ActivityIndicator] when isLoading prop is true', async () => {
    const comp = renderComponent({
      isLoading: true,
    });
    const activityIndicator = await queryActivityIndicator(comp);

    expect(activityIndicator).not.toBeNull();
  });

  it('should render [ActivityIndicator] with the value of indicatorColor prop', async () => {
    const SAMPLE_COLOR = 'red';
    const comp = renderComponent({
      isLoading: true,
      indicatorColor: SAMPLE_COLOR,
    });
    const activityIndicator = await queryActivityIndicator(comp);

    expect(activityIndicator.props.color).toEqual(SAMPLE_COLOR);
  });

  it('should render left icon when iconLeft prop is not empty', async () => {
    const SAMPLE_TEXT = 'sampleElementText';
    const sampleElement = (
      <Text>{SAMPLE_TEXT}</Text>
    );
    const comp = renderComponent({
      iconLeft: sampleElement,
    });
    const iconLeft = await queryIconLeft(comp);
    const icon = iconLeft.children[0] as ReactTestInstance;

    expect(iconLeft).not.toBeNull();
    expect(icon.children).toContain(SAMPLE_TEXT);
  });

  it('should not render left icon when isLoading prop is true', async () => {
    const SAMPLE_ELEMENT_TESTID = 'sampleElement';
    const sampleElement = (
      <Text testID={SAMPLE_ELEMENT_TESTID}>sampleElemenetText</Text>
    );
    const comp = renderComponent({
      iconLeft: sampleElement,
      isLoading: true,
    });
    const iconLeft = await queryIconLeft(comp);
    const icon = await comp.queryByTestId(SAMPLE_ELEMENT_TESTID);

    expect(iconLeft).toBeNull();
    expect(icon).toBeNull();
  });

  it('should render right icon when iconRight prop is not empty', async () => {
    const SAMPLE_TEXT = 'sampleElementText';
    const sampleElement = (
      <Text>{SAMPLE_TEXT}</Text>
    );
    const comp = renderComponent({
      iconRight: sampleElement,
    });
    const iconRight = await queryIconRight(comp);
    const icon = iconRight.children[0] as ReactTestInstance;

    expect(iconRight).not.toBeNull();
    expect(icon.children).toContain(SAMPLE_TEXT);
  });

  it('should not render right icon when isLoading prop is true', async () => {
    const SAMPLE_ELEMENT_TESTID = 'sampleElement';
    const sampleElement = (
      <Text testID={SAMPLE_ELEMENT_TESTID}>sampleElemenetText</Text>
    );
    const comp = renderComponent({
      iconRight: sampleElement,
      isLoading: true,
    });
    const iconRight = await queryIconRight(comp);
    const icon = await comp.queryByTestId(SAMPLE_ELEMENT_TESTID);

    expect(iconRight).toBeNull();
    expect(icon).toBeNull();
  });

  describe('interactions', () => {
    it('should call onClick on pressed', () => {
      const onClick = jest.fn();
      const comp = renderComponent({
        onClick,
      });
      const button = getButton(comp);

      fireEvent.press(button);
      expect(onClick).toHaveBeenCalled();
    });
  });
});
