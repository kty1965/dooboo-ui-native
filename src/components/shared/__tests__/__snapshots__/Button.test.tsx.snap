// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Button] renders without crashing 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="button"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "#0000ff",
            "borderRadius": 4,
            "borderWidth": 2,
            "flexDirection": "row",
            "height": 52,
            "justifyContent": "center",
            "width": 320,
          },
          Object {},
        ]
      }
      theme={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#0000ff",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "left": 0,
              "position": "absolute",
            },
          ]
        }
        testID="iconLeft"
      />
      <Text
        style={
          Array [
            Object {
              "color": "#069ccd",
              "fontSize": 14,
            },
            Object {},
          ]
        }
        testID="text"
        theme={
          Object {
            "fontColor": "#069ccd",
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "position": "absolute",
              "right": 0,
            },
          ]
        }
        testID="iconRight"
      />
    </View>
  </View>
</View>
`;

exports[`[Button] renders without crashing when children is not empty 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="button"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "#0000ff",
            "borderRadius": 4,
            "borderWidth": 2,
            "flexDirection": "row",
            "height": 52,
            "justifyContent": "center",
            "width": 320,
          },
          Object {},
        ]
      }
      theme={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#0000ff",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "left": 0,
              "position": "absolute",
            },
          ]
        }
        testID="iconLeft"
      />
      <Text>
        Sample text
      </Text>
      <View
        style={
          Array [
            Object {
              "position": "absolute",
              "right": 0,
            },
          ]
        }
        testID="iconRight"
      />
    </View>
  </View>
</View>
`;
